Building an API pt. 6 - lesson 7: POST /books Test (#165)
Ada-Developers-Academy/core-unit-2: api-6-testing/post-books-test.md

Reviewer: Super minor, but for the table row with the test declaration (around line 40), the wrapping lands so that the final `):` ends up on its own line looking very sad. Could we add a zdero width space after the open paren so that the wrap can happen there instead?

Also code quotes pytest

```md
|`def test_create_one_book(`&ZeroWidthSpace;`client):` | Continuing our best `pytest` practices, this test should start with the name `test_`, and it should describe the nature of this test.|
```

Our goals for this lesson are to:

> - Use fixtures to create test data

Reviewer: I was thinking to completely remove this one about fixtures, since this lesson doesn't use fixtures to create any test data this time. If we still want multiple goals, maybe this could be

```suggestion
- Use fixtures to access the test client to send a POST request
```

But that seems wordy to me.

Author: Ahh, I get what you're saying, I think it's fine to only have one goal and remove the fixture related one since we aren't writing new ones or doing anything new with them.

|`assert response_body == {...}}`| Checks for the expected key/value pairs in the response body|

As always, we should run `pytest` and ensure that the `POST` test passes and no other tests have been affected by our changes.

![Screenshot of pytest test result: 3 tests in tests/test_book_routes.py passed](../assets/api-6-testing/api-6-testing_post_test_success.png)  
_Fig. Test output reporting 3 tests in `tests/test_book_routes.py` passed. ([Full size image](../assets/api-6-testing/api-6-testing_post_test_success.png))_

## Check for Understanding

<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: d27c63d0-8dd3-4c09-ab95-e23e9d128aea
* title: POST /books Test
##### !question

Which of the statements below is true about why the test above, `test_create_one_book`, only needs the `client` fixture dependency? Select one option below.

##### !end-question
##### !options

a| The `client` fixture also creates data used by the test `test_create_one_book`.
b| `test_create_one_book` doesn't need the `client` fixture.
c| `test_create_one_book` doesn't need to retrieve a pre-existing record, so it does not need a fixture that saves Book records.

> d| The `client` fixture needs to used for every test using the `pytest` framework.

Author: ```suggestion
d| The `client` fixture needs to be used for every test using the `pytest` framework.
```

> |`client.post("/books", json=...`|Sends an `POST` request to `/books` with the JSON request body passed in as the `json` keyword argument| 

Reviewer: Rewords to emphasize the distinction between the actual dict and the JSON it will become
```suggestion
|`client.post("/books", json=...`|Sends a `POST` request to `/books`, with the `dict` to be used as the JSON request body passed in using the `json` keyword argument| 
```

> | `response_body = response.get_json()`| Get the JSON response body with `response.get_json()`|

Reviewer: Rewords to emphasize the distinction between the JSON the response contained and the value we get in python
```suggestion
| `response_body = response.get_json()`| Gets the JSON response body as a Python value. Since the create logic returns a `dict`, we should anticipate `response_body` to receive a `dict` result here.|
```

> |`assert response.status_code == 201`|Check for the expected status code|

Reviewer: follows tense established in "sends a post" (combining with the next suggestion gave a very strange git diff)
```suggestion
|`assert response.status_code == 201`|Checks for the expected status code|
```

> |`assert response_body == {...}}`| Check for the expected key/value pairs in the response body|

Reviewer: follows tense established in "sends a post"
```suggestion
|`assert response_body == {...}}`| Checks for the expected key/value pairs in the response body|
```

As always, we should run `pytest` and ensure that the `POST` test passes and no other tests have been affected by our changes.

![Screenshot of pytest test result: 3 tests in tests/test_book_routes.py passed](../assets/api-6-testing/api-6-testing_post_test_success.png)  

> _Fig. Test output reporting 3 tests in `tests/test_book_routes.py` passed. ([Full size image](../assets/api-6-testing/api-6-testing_post_test_success.png))_

Reviewer: Thank you!

## Check for Understanding

<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: d27c63d0-8dd3-4c09-ab95-e23e9d128aea
* title: POST /books Test
##### !question

Which of the statements below is true about why the test above, `test_create_one_book`, only needs the `client` fixture dependency? Select one option below.

##### !end-question
##### !options

a| The `client` fixture also creates data used by the test `test_create_one_book`.
b| `test_create_one_book` doesn't need the `client` fixture.
c| `test_create_one_book` doesn't need to retrieve a pre-existing record, so it does not need a fixture that saves Book records.
d| The `client` fixture needs to used for every test using the `pytest` framework.

##### !end-options
##### !answer

c|

##### !end-answer
##### !hint

What does the client fixture do for us? When do we need fixtures which create data and why?

##### !end-hint
##### !explanation

1. False: The `client` fixture does not create any `Book` records
2. False: The `client` fixture allows us to make requests to our API, so we must include it as a dependency.

> 3. True: We only need a fixture which creates a `Book` record before the test runs if the test depends on there being data in the database.

Reviewer: uses "that" for a restrictive clause
```suggestion
3. True: We only need a fixture that creates a `Book` record before the test runs if the test depends on there being data in the database.
```
